src/ServiceControl/Bootstrapper.cs

```diff
@@ -77,12 +81,25 @@ private void Initialize()
             containerBuilder.RegisterInstance(loggingSettings);
             containerBuilder.RegisterInstance(settings);
             containerBuilder.RegisterInstance(notifier).ExternallyOwned();
-            containerBuilder.RegisterInstance(timeKeeper).ExternallyOwned();
+            containerBuilder.RegisterInstance(timeKeeper).AsSelf().AsImplementedInterfaces().ExternallyOwned();
             containerBuilder.RegisterType<SubscribeToOwnEvents>().PropertiesAutowired().SingleInstance();
             containerBuilder.RegisterInstance(documentStore).As<IDocumentStore>().ExternallyOwned();
+            containerBuilder.RegisterType<EndpointUptimeInformationPersister>().AsImplementedInterfaces().SingleInstance();
             containerBuilder.Register(c => HttpClientFactory);
             containerBuilder.RegisterModule<ApisModule>();
 
+            foreach (var component in components)
+            {
+                component.RegisterDependency(containerBuilder);
+
+                var parts = component.CreateParts();
+
+                foreach (var part in parts)
+                {
+                    containerBuilder.RegisterInstance(part).ExternallyOwned().AsImplementedInterfaces().AsSelf();
```

Comment:
Maybe we should invert this so that the creation of the parts has access to the `containerBuilder` in the event that they want to register their own dependencies.

I'm thinking in the case where we want to persist using Sql. Raven will be provided by default, but the component (or piece or part or whatever) could register a SqlConnection factory in the containerBuilder so that it's internal classes will be provided with an instance when they ask for one.


E.g.:


```CSharp
var parts = component.CreateParts(containerBuilder);

...

IEnumerable<objecT>CreateParts(ContainerBuilder builder)
{
    builder.RegisterInstance(sqlPersistenceThing).ExternallyOwned();

    var uptimeModule = new UptimeApiModule(monitoring, builder); // here they can register too
    builder.RegisterInstance(uptimeModule).ExternallyOwned().AsImplementedInterfaces().AsSelf();
    yield return uptimeModule;

    var heartbeatFailureDetected = new HeartbeatFailureDetector(monitoring);
    builder.RegisterInstance(heartbeatFailureDetected).ExternallyOwned().AsImplementedInterfaces().AsSelf();
    yield return heartbeatFailureDetected;
}
```

Reply:
I've replied to @mikeminutillo 's similar suggestion here https://github.com/Particular/ServiceControl/pull/1112/files#r175994712. TL;DR I believe we should not start with a container. The new stuff should be based on a simple model we use in the core where a root (feature/component) creates instances of shared components (e.g. `EndpointInstanceMonitoring`) and parts (that do job e.g. behaviors)